Homework25

rramoscode opened the pull request.

rramoscode committed:
Homework16 Test Registration Navigation

-Create a new file, Homework16.java in IntelliJ IDEA.
-Create a test case, registrationNavigation() using @Test annotation.
-Using Selenium, navigate to "https://bbb.testpro.io/".
-Click the Registration link.
-Verify that you are redirected to the Registration page using the Assert method.

rramoscode committed:
Test Case  ( addSongToPlaylist()  ) using @Test annotation and the helper/reusable methods we created.

steps completed:
- Navigate to "https://bbb.testpro.io/".
- Log in with your credentials.
-  Search for a song (choose any song of your choice).
- Click View All to display the search results.
- Click the first song in the search results.
- Click ADD TO...
- Choose the playlist to add it to, (we can create a new one with a unique name).
- Verify that the notification message appears.
- Verify that notification message has the text, "Added 1 song into {your playlist}".

rramoscode committed:
Update BaseTest.java

updated note for personal reference

rramoscode committed:
Homework18 created Tc for playSong and new .xml file with method to run

- Create a test case, playSong() using @Test annotation.
- Use the helper/reusable methods we created.
- Create a new config XML file for the test case.
- Specify the method to be ran in the XML file.
- Navigate to "https://bbb.testpro.io/".
- Log in with your credentials.
- Click «Play next song» (media player controls), then the Play button, to play a song.
- Validate that a song is playing by verifying if the sound bar is displayed.

rramoscode committed:
testing no changes made

review code option no real changes made to of previous code

rramoscode committed:
Parameter Annotation in TestNG

 - Navigate to "https://bbb.testpro.io/".
- Log in with your credentials.
- Select the playlist you want to delete.
- Click the red button "x PLAYLIST" on the top right part of the page, and click on it.
- Verify that the confirmation notification displayed has the text, "Deleted playlist  ".

rramoscode committed:
no changes made

there was no changes made , was looking at different option for waits

rramoscode committed:
refactored code to have Explicit Wait

refactored code to have Explicit Waits, instead for Thread.sleep.
- Refactor your code - implement Explicit waits.
- Remove all implementations of Thread.sleep.
- Make sure our codes run without errors.

rramoscode committed:
Explicit wait added to all methods

Added explicit wait to all methods for HW 16-18 although not all methods may be needed for HW20.

rramoscode committed:
recommended updates made

removed   throws interruptedExceptions  and unused method for non-empty playlist

rramoscode committed:
rename playlist code

- Create a new file Homework21.java for the renamePlaylist script
- Used Action Class methods, create/refactor the test case renamePlaylist().
- Implemented Explicit Waits.
- Remove all implementations of Thread.sleep. (used only during debugging)

script changes rrPlaylist3 to rrPlaylist4

rramoscode committed:
refactor Pull member

updated Homework21 and refactored helper methods using pull member to move helper methods to BaseTest

rramoscode committed:
Update BaseTest.java

update made due to noticed the helper methods had protected void... instead of public void

rramoscode committed:
Update file-system.probe

reran Homework21 no issues

rramoscode committed:
update to Assert

updated assertion to include string name

rramoscode committed:
update to Assertion

practice change for different looks

rramoscode committed:
homework22 refactor and implement code to POM

Refactored the code, implemented and created Page Objects.

rramoscode committed:
Update BasePage.java

per feedback
updated variable to protected to prevent non-related classes from trying to use it.

rramoscode committed:
Update LoginPage.java

implemented encapsulation to the Loginpage

rramoscode committed:
Homework23 selenium page Factory

- Refactored the code and implemented Page Factory by using the initElements() method of Page Factory class.
-  Used @FindBy annotation for declaring Web Elements.

rramoscode committed:
Update file-system.probe

no change made

rramoscode committed:
Update file-system.probe

no changes made

rramoscode committed:
suggested changes made

made suggested changes based on feedback

rramoscode committed:
Update BaseTest.java

rramoscode committed:
homework24 using standalone option locally

- Set up  Selenium Grid (Standalone option) locally. Chrome/ firefox browser etc..
- Implement Browser Factory method as we had in the class, so I could run Chrome browser using your local Grid.

test changes rrPlaylist3 to rrPlaylist4

rramoscode committed:
Update BaseTest.java

updated  edge information (educational purposes only)

rramoscode committed:
Update BaseTest.java

no changes to code made

rramoscode committed:
Homework25 cloud execution

- created new branch Homework25
- added the cloud execution capability to the framework for use with  Lamdatest.com
- ran @test renamePlaylist in the cloud

Berichicko commented:
Please, add more tests in TestNG and run

For example

![image](https://github.com/Test-Pro-Bootcamp/A45-QA-Automation-Framework/assets/104761482/2642e178-6ecc-4100-a60e-4caa4917c593)

rramoscode committed:
aded to to TestNG

added additional @test to TestNG and ran a test through cloud LambdaTest  all passed.
- LoginTest - sucessfulLonginTest - Passed
- Homework20 removeplaylist (rrPlaylist2) - Passed
-Homework24 renamePlaylist - renames rrPlaylist3 to rrPlaylist4 - Passed

rramoscode commented:
added additional test as requested, will resubmit for review.

Berichicko commented:
Please finish the homework!You got this problem, please correct it. 

![image](https://github.com/Test-Pro-Bootcamp/A45-QA-Automation-Framework/assets/104761482/70586743-761a-42f8-9761-15e3c1f6721e)

![image](https://github.com/Test-Pro-Bootcamp/A45-QA-Automation-Framework/assets/104761482/b97b93ee-02c5-4fc3-a456-37d9722418d5)

Dear Randy Ramos

I'm trying to help, so I wrote a code that can solve your problems.

1)  **Step 1**

_Let’s execute TestNG classes in parallel using TestNG xml as below:_

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="TestNG" parallel="methods" thread-count="4">
    <test name="Example Test" preserve-order="false">
        <parameter name="baseURL" value="https://bbb.testpro.io/"/>
        <classes>
            <class name="PlaylistTests"/>
            <class name="LoginTests"/>
        </classes>
    </test>
</suite>
```


2) **Step 2** 

_Let’s declare a static WebDriver as ThreadLocal. I have created simple methods to set, get and remove driver._

```
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.edge.EdgeDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.RemoteWebDriver;
import org.testng.annotations.*;

import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.time.Duration;

public class BaseTest {

    private static final ThreadLocal<WebDriver> THREAD_LOCAL = new ThreadLocal<>();
    private WebDriver driver;

    public static WebDriver getThreadLocal() {
        return THREAD_LOCAL.get();
    }

    @BeforeMethod
    @Parameters({"baseURL"})
    public void setUpBrowser(@Optional String baseURL) throws MalformedURLException {
        THREAD_LOCAL.set(pickBrowser(System.getProperty("browser")));
        THREAD_LOCAL.get().manage().timeouts().implicitlyWait(Duration.ofSeconds(5));
        THREAD_LOCAL.get().manage().window().maximize();
        THREAD_LOCAL.get().manage().deleteAllCookies();
        getThreadLocal().get(baseURL);
        System.out.println(
                "Browser setup by Thread " + Thread.currentThread().getId() + " and Driver reference is : " + getThreadLocal());

    }

    public WebDriver lambdaTest() throws MalformedURLException {
        String username = "";
        String authkey = "";
        String hub = "@hub.lambdatest.com/wd/hub";
        DesiredCapabilities caps = new DesiredCapabilities();
        caps.setCapability("platform", "Windows 10");
        caps.setCapability("browserName", "Chrome");
        caps.setCapability("version", "110.0");
        caps.setCapability("resolution", "1024x768");
        caps.setCapability("build", "TestNG With Java");
        caps.setCapability("name", this.getClass().getName());
        caps.setCapability("plugin", "git-testng");
        return new RemoteWebDriver(new URL("https://" + username + ":" + authkey + hub), caps);
    }

    public WebDriver pickBrowser(String browser) throws MalformedURLException {
        DesiredCapabilities capabilities = new DesiredCapabilities();
        String gridURL = "http://10.2.127.17:4444";

        switch (browser) {
            case "firefox":
                WebDriverManager.firefoxdriver().setup();
                FirefoxOptions optionsFirefox = new FirefoxOptions();
                optionsFirefox.addArguments("-private");
                return driver = new FirefoxDriver(optionsFirefox);
            case "edge":
                WebDriverManager.edgedriver().setup();
                return driver = new EdgeDriver();
            case "grid-firefox":
                capabilities.setCapability("browserName", "firefox");
                return driver = new RemoteWebDriver(URI.create(gridURL).toURL(), capabilities);
            case "grid-edge":
                capabilities.setCapability("browserName", "edge");
                return driver = new RemoteWebDriver(URI.create(gridURL).toURL(), capabilities);
            case "grid-chrome":
                capabilities.setCapability("browserName", "chrome");
                return driver = new RemoteWebDriver(URI.create(gridURL).toURL(), capabilities);
            case "cloud":
                return lambdaTest();
            default:
                WebDriverManager.chromedriver().setup();
                ChromeOptions optionsChrome = new ChromeOptions();
                optionsChrome.addArguments("--disable-notifications","--remote-allow-origins=*", "--incognito","--start-maximized");
                optionsChrome.setExperimentalOption("excludeSwitches", new String[]{"enable-automation"});
                return driver = new ChromeDriver(optionsChrome);
        }
    }

    @AfterMethod
    public void tearDown() {
        THREAD_LOCAL.get().close();
        THREAD_LOCAL.remove();
    }
}
```


3) **Step 3**

```
import org.testng.Assert;
import org.testng.annotations.Test;
import pages.HomePage;
import pages.LoginPage;

public class LoginTests extends BaseTest {
    @Test
    public static void loginEmptyEmailPasswordTest() {
        LoginPage loginPage = new LoginPage(getThreadLocal());

        loginPage.provideEmail("")
                .providePassword("te$t$tudent")
                .clickSubmitBtn();

        Assert.assertTrue(loginPage.getRegistrationLink().isDisplayed());
    }

    @Test
    public static void loginWrongPasswordTest() {
        LoginPage loginPage = new LoginPage(getThreadLocal());

        loginPage.provideEmail("demo@class.com")
                .providePassword("te$t123")
                .clickSubmitBtn();

        Assert.assertTrue(loginPage.getRegistrationLink().isDisplayed());
    }

    @Test
    public static void loginEmptyPasswordTest() {
        LoginPage loginPage = new LoginPage(getThreadLocal());

        loginPage.provideEmail("demo@class.com")
                .providePassword("")
                .clickSubmitBtn();

        Assert.assertTrue(loginPage.getRegistrationLink().isDisplayed());
    }

    @Test
    public static void loginWrongEmailTest() {
        LoginPage loginPage = new LoginPage(getThreadLocal());

        loginPage.provideEmail("demo@class.com")
                .providePassword("te$t$tudent")
                .clickSubmitBtn();

        Assert.assertTrue(loginPage.getRegistrationLink().isDisplayed());
    }

    @Test
    public void loginSucceedTest() {
        LoginPage loginPage = new LoginPage(getThreadLocal());
        HomePage homePage = new HomePage(getThreadLocal());

        loginPage.provideLoginSucceed();

        Assert.assertTrue(homePage.getUserAvatar().isDisplayed());
    }
}
```

```
import org.testng.Assert;
import org.testng.annotations.Test;

import pages.HomePage;
import pages.LoginPage;

public class PlaylistTests extends BaseTest {

    @Test
    public void deletePlaylistTest() {

        LoginPage loginPage = new LoginPage(getThreadLocal());
        HomePage homePage = new HomePage(getThreadLocal());

        loginPage.provideLoginSucceed();
        homePage.chooseFirstPlaylist()
                .deletePlaylist();

        Assert.assertTrue(homePage.notificationText());
    }
}
```

3) **Step 4**

```
plugins {
    id 'java'
}

group 'org.example'
version '1.0-SNAPSHOT'

repositories {
    mavenCentral()
}

dependencies {
    testImplementation 'org.testng:testng:7.6.1'
    implementation 'org.seleniumhq.selenium:selenium-java:4.5.0'
    testImplementation 'io.github.bonigarcia:webdrivermanager:5.3.1'
}

test {
    useTestNG() {
        suites "src/test/resources/TestNG.xml"
        systemProperty "browser", System.getProperty("browser")
    }
}
```

3) **Step 5**

**Running Tests
To run tests, run the following command**

`gradle clean test`

![image](https://github.com/Svyeta0210/A45-QA-Automation-Framework/assets/104761482/4bcd07e2-73b7-42a9-88ad-b9d3e13b6466)


**For example, download** https://drive.google.com/file/d/1kYRPi5xvwHRMLglSxhkBLJKUdX3Ofy5s/view?usp=share_link

Berichicko commented:
**Here's an example code that can be run on a local machine with 4 threads:**

_Download_

1) Code with Russian comments https://drive.google.com/file/d/1dQjnroe0pIiIqll_dT8rTwgAO5cekj4W/view?usp=share_link


2) Code with English comments https://drive.google.com/file/d/1H6MUQcisY_XQ2kn47AhrGU5TBVxli1ae/view?usp=share_link


**To run the tests, open the README file in IntelliJ IDEA**


![image](https://github.com/Test-Pro-Bootcamp/A45-QA-Automation-Framework/assets/104761482/3c7532eb-65f0-4854-823b-be41ba43622a)

**click**  ![image](https://github.com/Test-Pro-Bootcamp/A45-QA-Automation-Framework/assets/104761482/7313f1bd-6aba-4a26-8f83-dd510a432c8c)

_PS:  To unzip a ZIP file on a Mac_ https://www.winzip.com/en/learn/file-formats/zip/mac/

**Do I have to run tests in Lambdatest or on my local machine?**
_It depends on your needs. Lambdatest is for cross-browser and device testing, but running tests locally is fine_

rramoscode commented:
test working when using - gradle clean test - will resubmit for review
test performed:
successful login test
delete rrPlaylist2
rename rrPlaylist3 to rrPlaylist4

![Screen Shot 2023-05-12 at 3 13 59 PM](https://github.com/Test-Pro-Bootcamp/A45-QA-Automation-Framework/assets/25727390/2456ce2e-14bb-48f5-b5c6-2179ae859cf0)

rramoscode committed:
update to correct errors

test working when using - gradle clean test - will resubmit for review
test performed:
successful login test
delete rrPlaylist2
rename rrPlaylist3 to rrPlaylist4

Berichicko commented:
Hello 
Your homework 25 is checked
**You’re doing great! Keep it up! Very fine!** 
💖Best wishes, the team HW (Alex,  aliaksandr@testpro.io)!💖

![804deb64dc406a20a75da05d91a1a69e](https://user-images.githubusercontent.com/104761482/235961479-9afb2841-b657-46c9-99f8-2402f933b843.jpg)
